File: C++/394. Decode String.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<string> strStack;
        stack<int> nums;
        string ans;
        int num = 0;
        for(char c : s) {
            if(c>='0' && c<='9') {
                num = num*10 + (c-'0');                              
            }
            else if(c == '[') {
                strStack.push(ans);
                nums.push(num);
                ans = "";
                num = 0;
            }
            else if(c == ']') {
                string temp = ans;
                for(int i = 0; i < nums.top()-1; ++i) {
                    ans += temp;
                }
                ans = strStack.top() + ans;
                strStack.pop();
                nums.pop();
            }
            else{
                ans.push_back(c);
            }
        }
        return ans;
    }
};

File: Java/394. Decode String.java
class Solution {
    public String decodeString(String s) {
        Stack<StringBuilder> strStack = new Stack<>();
        Stack<Integer> nums = new Stack<>();
        StringBuilder ans = new StringBuilder();
        int num = 0;
        for(char c:s.toCharArray()) {
            if(c>='0' && c<='9') {
                num = num*10 + (c-'0');                              
            }
            else if(c == '[') {
                strStack.push(ans);
                nums.push(num);
                ans=new StringBuilder();
                num = 0;
            }
            else if(c == ']') {
                StringBuilder temp=ans;
                int i = nums.pop();
                ans = strStack.pop();
                while(i-->0) {
                    ans.append(temp);
                }
            }
            else{
                ans.append(c);
            }
        }
        return ans.toString();
    }
}
